btree.go

```diff
@@ -0,0 +1,346 @@
+	bucketSize int
+	splitKey   ngram
+}
+
+func (n *innerNode) sizeBytes() int {
+	return len(n.keys)*ngramEncoding + len(n.children)*int(interfaceBytes)
+}
+
+func (n *leaf) sizeBytes() int {
+	return 4 * 8
+}
+
+func (n *leaf) insert(ng ngram, opts btreeOpts) {
+	n.bucketSize++
+
+	if n.bucketSize == (opts.bucketSize/2)+1 {
+		n.splitKey = ng
+	}
+}
+
+func (n *innerNode) insert(ng ngram, opts btreeOpts) {
+	insertAt := func(i int) {
+		// Invariant: Nodes always have a free slot.
+		//
+		// We split full nodes on the the way down to the leaf. This has the
+		// advantage that inserts are handled in a single pass.
+		if leftNode, rightNode, newKey, ok := n.children[i].maybeSplit(opts); ok {
+			n.keys = append(n.keys[0:i], append([]ngram{newKey}, n.keys[i:]...)...)
+			n.children = append(n.children[0:i], append([]node{leftNode, rightNode}, n.children[i+1:]...)...)
+
+			// A split might shift the target index by 1.
+			if ng >= n.keys[i] {
+				i++
+			}
+		}
+		n.children[i].insert(ng, opts)
+	}
+
+	for i, k := range n.keys {
+		if ng < k {
+			insertAt(i)
+			return
+		}
+	}
+	insertAt(len(n.children) - 1)
+}
+
+// See btree.find
+func (n *innerNode) find(ng ngram) (int, int) {
+	for i, k := range n.keys {
+		if ng < k {
+			return n.children[i].find(ng)
+		}
+	}
+	return n.children[len(n.children)-1].find(ng)
+}
+
+// See btree.find
+func (n *leaf) find(ng ngram) (int, int) {
+	return int(n.bucketIndex), int(n.postingIndexOffset)
+}
+
+func (n *leaf) maybeSplit(opts btreeOpts) (left node, right node, newKey ngram, ok bool) {
+	if n.bucketSize < opts.bucketSize {
+		return
+	}
+	return &leaf{bucketSize: opts.bucketSize / 2},
+		&leaf{bucketSize: opts.bucketSize / 2},
+		n.splitKey,
+		true
+}
+
+func (n *innerNode) maybeSplit(opts btreeOpts) (left node, right node, newKey ngram, ok bool) {
+	if len(n.children) < 2*opts.v {
+		return
+	}
+	return &innerNode{
+			keys:     append(make([]ngram, 0, opts.v-1), n.keys[0:opts.v-1]...),
+			children: append(make([]node, 0, opts.v), n.children[:opts.v]...)},
+		&innerNode{
+			keys:     append(make([]ngram, 0, (2*opts.v)-1), n.keys[opts.v:]...),
+			children: append(make([]node, 0, 2*opts.v), n.children[opts.v:]...)},
+		n.keys[opts.v-1],
+		true
+}
+
+func (n *leaf) visit(f func(n node)) {
+	f(n)
+	return
+}
+
+func (n *innerNode) visit(f func(n node)) {
+	f(n)
+	for _, child := range n.children {
+		child.visit(f)
+	}
+}
+
+func (bt *btree) String() string {
+	s := ""
+	s += fmt.Sprintf("%+v", bt.opts)
+	bt.root.visit(func(n node) {
+		switch nd := n.(type) {
+		case *leaf:
+			return
+		case *innerNode:
+			s += fmt.Sprintf("[")
+			for _, key := range nd.keys {
+				s += fmt.Sprintf("%d,", key)
+			}
+			s = s[:len(s)-1] // remove coma
+			s += fmt.Sprintf("]")
+
+		}
+	})
+	return s
+}
+
+type btreeIndex struct {
+	bt *btree
+
+	// We need the index to read buckets into memory.
+	file IndexFile
+
+	bucketOffsets        []uint32
+	bucketSentinelOffset uint32
```

Comment:
minor: sentinels usually are stored in the slice to make this easier. IE you have N buckets, but bucketOffsets has size N+1 with the last value being this special value. That should simplify code below?

Reply:
👍 I copy&pasted this idea from binarySearchNgram ;-)

Reply:
hehe, in that case though we didn't want to mutate the slice to add the sentinal. In this case we do construct the full slice here so we can mutate it and just push the value on.

Reply:
ok, ok :-)